[3.x] Fix scale sensitivity for 3D objects.

lentsius-bark opened the pull request:
Scaling 3D objects has been broken for a long time, it works well **only** when scale values stay around their original ones. Meaning that if you try scaling an objects that already at scale 30, the scaling function becomes unusable. Simple as that.

**Current godot implementation vs. this PR:**

https://user-images.githubusercontent.com/38346332/133256881-79c4eb1c-9e4d-4ee0-8f3c-f5645b0d1046.mp4

This PR also divides the scale factor by the distance of the mouse click from the object. This is neat as it gives users the ability to choose scaling sensitivity in a very simple way: "The further from the object's center you mouse is, the slower it scales and vice versa"

**A video showing the variable scaling sensitivity:**

https://user-images.githubusercontent.com/38346332/133257142-bbb31dcf-c1e2-4f94-9b8b-ec501f0bef58.mp4

I'm new to contributing and a novice at C++, I welcome all feedback. I heard the codebase for specifically this has changed loads for 4.0, chances are it's already fixed if not I could try my hand at porting it?

JFonS commented on editor/plugins/spatial_editor_plugin.cpp:
```diff
@@ -1647,6 +1647,9 @@ void SpatialEditorViewport::_sinput(const Ref<InputEvent> &p_event) {
 							Transform t;
 							Vector3 local_scale;
 
+							float click_distance = click.distance_to(_edit.center);
+							motion = ((motion / original.basis.get_scale()) / (Vector3(click_distance, click_distance, click_distance) / original.basis.get_scale()));
```
The divisions by `original.basis.get_scale()` are redundant since they happen on both sides of the larger division. The code can be simplified down to this:
```suggestion
							motion /= click.distance_to(_edit.center);
```

JFonS reviewed (changes requested):
Makes sense usability-wise. The way the scaling factor is computed is a bit convoluted, I left a change suggestion in a comment. 

This also needs to be changed in `master`. Simply applying the same scaling factor to `motion` right before this piece of code should do it: https://github.com/godotengine/godot/blob/c8c119940264cd1d6d20ee88c03ee20f376dafc9/editor/plugins/node_3d_editor_plugin.cpp#L1861-L1874

lentsius-bark commented:
I'll be online later this evening and work the changes in then, thanks for the suggestions! Loving the simplicity.

akien-mga commented:
Aside from the code simplification, two more things worth considering (I didn't assess them myself):
- Is this the "best" location to perform this operation? `motion` is calculated further up.
- Related, we're in a `for` loop for all the selection - was this tested with multiple scaled objects selected, with different scales? It looks like it will mutate the same `motion` each time so I suspect you'll see problems.
- Related, further up some calculations are done for when snapping is enabled. Was this tested with snapping on, and if so does it work as expected?

lentsius-bark committed:
Fix issues with scaling 3D Objects

When scaling 3D objects the distance form them is not considered. Allowing for finer contorl. Overscaled objects no longer break the gizmo.

lentsius-bark commented:
@JFonS I implemented your suggestions. Thanks for pointing this out, really good stuff.
@akien-mga I moved the code higher, right after where motion is first defined so that it'd work robustly. 
I tried scaling multiple objects as well as scale snap and it worked as expected.

I'll rebase and squash commits when this is all resolved :)

lentsius-bark commented:
a video with the scaling of a group as well as snap scale

https://user-images.githubusercontent.com/38346332/133335527-7f4b97c5-d7bf-4b74-8686-49a9810eb2a0.mp4

JFonS reviewed (approved):
Looks good. I approve the PR, but it would need a `master` version before being merged :)

lentsius-bark commented:
@JFonS https://github.com/godotengine/godot/pull/52685

Should I rebase and squash commits on this one?

JFonS commented:
@lentsius-bark yup, this one needs the commits squashed.

lentsius-bark commented:
@JFonS will do, thanks!

lentsius-bark commented:
@JFonS done.

akien-mga commented:
Thanks!
